framework/auth/views.py

```diff
@@ -202,92 +202,160 @@ def forgot_password_post(auth, **kwargs):
-                                   'resend email.</a>', trust=False)
 
-    if next_url and must_login_warning:
-        status.push_status_message(language.MUST_LOGIN, trust=False)
+@collect_auth
+def auth_register(auth):
+    """
+    View for OSF register. Land on the register page, redirect or go to `auth_logout`
+    depending on `data` returned by `login_and_register_handler`.
+
+    :param auth: the auth context
+    :return: land, redirect or `auth_logout`
+    :raise: http.BAD_REQUEST
+    """
+
+    context = {}
+    # a target campaign in `auth.campaigns`
+    campaign = request.args.get('campaign')
+    # the service url for CAS login or redirect url for OSF
+    service_url = request.args.get('next')
+    # used only for `claim_user_registered`
+    logout = request.args.get('logout')
+
+    data = login_and_register_handler(auth, login=False, campaign=campaign, service_url=service_url, logout=logout)
+
+    # land on register page
+    if data['status_code'] == http.OK:
+        if data['must_login_warning']:
+            status.push_status_message(language.MUST_LOGIN, trust=False)
+        context['non_institution_login_url'] = cas.get_login_url(data['service_url'])
+        context['institution_login_url'] = cas.get_login_url(data['service_url'], campaign='institution')
+        context['campaign'] = data['campaign']
+        return context, http.OK
+    # redirect to url
+    elif data['status_code'] == http.FOUND:
+        return redirect(data['redirect_url'])
+    # go to other views
+    elif data['status_code'] == 'auth_logout':
+        return auth_logout(redirect_url=data['redirect_url'])
+
+    raise HTTPError(http.BAD_REQUEST)
+
 
-    # set login_url to form action, upon successful authentication specifically w/o logout=True,
-    # allows for next to be followed or a redirect to the dashboard.
-    redirect_url = web_url_for('auth_login', next=next_url, _absolute=True)
+@collect_auth
+def auth_login(auth):
+    """
+    View (no template) for OSF Login.
+    Redirect user based on `data` returned from `login_and_register_handler`.
 
-    data = {}
-    if campaign and campaign in campaigns.CAMPAIGNS:
-        if (campaign == 'institution' and settings.ENABLE_INSTITUTIONS) or campaign != 'institution':
-            data['campaign'] = campaign
-    data['login_url'] = cas.get_login_url(redirect_url)
-    data['institution_redirect'] = cas.get_institution_target(redirect_url)
-    data['redirect_url'] = next_url
-    data['sign_up'] = request.args.get('sign_up', False)
-    data['existing_user'] = request.args.get('existing_user', None)
+    :param auth: the auth context
+    :return: redirects
+    """
 
-    return data, http.OK
+    campaign = request.args.get('campaign')
+    data = login_and_register_handler(auth, login=True, campaign=campaign)
+    if data['status_code'] == http.FOUND:
+        return redirect(data['redirect_url'])
+    # TODO: is `http.FOUND` guaranteed?
```

Comment:
Unless it's an invalid campaign, which would raise an error, it appears to be guaranteed as long as `login=True` is passed in.